### ISSUE TITLE
digital vomit with latest MESA

### ISSUE THREAD

**Issue by dukenukemx** [RobertBeckebans/RBDOOM-3-BFG#386]
I'm running Mint 18.2 with RBDOOM3BFG installed from Ubuntu repository, and I had no problem running the game with my Radeon HD 7850.  But when I installed Padoka PPA for latest AMD RadeonSi drivers I get corrupted screen that's incomprehensible.  Essentially digital vomit.  I also tried compiling the latest version from this Github with the same results.

**Comment by BielBdeLuna** [comment]
capture the screen and post the result here to see what sort of problem you might have

**Comment by dukenukemx** [comment]
Here is what it looks like.  I've tried games like Portal 2, Borderlands 2, and Dolphin and they all render correctly.  
![doom3bfg](https://user-images.githubusercontent.com/31163475/29541369-954b34b6-86a1-11e7-9825-4305f71498b5.png)

**Comment by spannerman79** [comment]
That looks like what used to happen with an older nvidia driver blob and Chrome (unrelated to this project).

> But when I installed Padoka PPA for latest AMD RadeonSi drivers

RadeonSI/Gallium3D drivers are too problematic.  The same can be said about using nouveau - for which I will never do for gaming.

No need to use a PPA - http://support.amd.com/en-us/download/linux

**Comment by dukenukemx** [comment]
The RadeonSi drivers are currently much faster than AMDGPU drivers, except for Vulkan performance.  Plus it gives me OpenGL4.5 support.  Also the glitching you're seeing is not from a web browser but from the game itself.  How the Google crap got in there is beyond me.

**Comment by DanielGibson** [comment]
Probably uninitialized memory (buffers) that still holding old data.
Unfortunately on Linux GPU drivers don't zero out the graphics memory before assigning it to a new process.

**Comment by dukenukemx** [comment]
A simple solution would be to remove the Padoka PPA, but then I downgrade my graphics driver.  So maybe this is a new bug introduced to the drivers since these are bleeding edge, but then why does nothing else have problems?

Also for the hell of it I tried "MESA_GL_VERSION_OVERRIDE=3.3 rbdoom3bfg" and it gave the error "ERROR: GL_ARB_multitexture not available".  So that's odd.  I tried with opengl4.5 with no difference.

**Comment by dukenukemx** [comment]
Have a laptop with a Radeon HD 6370 GPU and it hadn't been updated in a while, so I tried RBDOM3BFG and it worked fine with Padoka PPA.  Updated the system, which means updated drivers and I get the same digital vomit.  Which means a new update broke something.

**Comment by shmerl** [comment]
Here is what I get with latest Mesa master and AMD RX 480:

![doom3_bfg_garbled](https://user-images.githubusercontent.com/310927/29747228-69b035f4-8ac0-11e7-8df8-867c3cb49b91.jpg)

**Comment by shmerl** [comment]
I'm also getting this during run:

    0:2(10): error: GLSL 1.50 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES

I think it mixes up OpenGL version, and doesn't run in core profile.

**Comment by shmerl** [comment]
OK, I was able to work around it like this. I commented out all checks like:

    if( !glConfig.multitextureAvailable )
...
    if( !glConfig.textureCompressionAvailable )

And so on, which verify availablity of OpenGL extensions. Then I run the game like this:

    MESA_GL_VERSION_OVERRIDE=4.5 MESA_GLSL_VERSION_OVERRIDE=450 mesa_run.sh ./RBDoom3BFG

And it works beautifully, and even antialiasing works as it should.

**Comment by shmerl** [comment]
Yeah, something is quite off with graphics when using this method. Here is how flashlight lighting looks:

![doom3_flashlight_bug](https://user-images.githubusercontent.com/310927/29748148-00692966-8ade-11e7-9eee-382ecc1d0b18.jpg)

**Comment by dukenukemx** [comment]
I updated the title of this so that it isn't just about Padoka.  I've also tried Oibaf PPA with similar results.  So it seems that because RBDOOM3BFG uses compat profile instead of core profile that this is what causes the problem?  

Also is there an official fix or we need to comment out some if statements?

**Comment by shmerl** [comment]
I'm not sure if commenting out those checks is a proper fix. Developers can probably explain it better, but I suspect that the game relies on compat profile, and it's a problem. It really should be rewritten to use core profile only.

**Comment by shmerl** [comment]
Yep, extensions that the game log reports are from compat profile, not from core profile.

**Comment by dsalt** [comment]
#391 may be related – re-test with that applied and see what you get.

**Comment by shmerl** [comment]
I'll try it, but I don't think it will help incorrect initialization.

**Comment by BielBdeLuna** [comment]
in the image what is it the problem? is it the added brightness in the planar flashlight texture (the one simulating the light interaction with the medium)? if so, this is due that robert changed the gamma handling of textures for the new renderer, so transparent textures where designed for an incorrect gamma handling, so all transparent effects look either too bright or too faint ( like imp fireballs that look way too dark ) this is due the assets where made for that incorrect gamma math in the vanilla renderer.

**Comment by shmerl** [comment]
That image appears like some gray mesh, instead of actual light effect.

**Comment by BielBdeLuna** [comment]
I use Intel and it looks the same as your image, those problems where announced when Robert applied the new Gamma math, it's the assets. take a look how imps fireball look, and you can see any let's play to see how they should look like, the same problem the assets conflicting with the new Gamma math.

**Comment by shmerl** [comment]
I see. And there is no way to fix that?

**Comment by BielBdeLuna** [comment]
there are three ways, none of them is my favourite:

- go back to the old wrong gamma math ( big no no for me )
- rework the assets themselves ( way too much work )
- add some kind of token in the materials of those textures so their gamma is handled specially or they are preprocessed in some way ( way too much work just for old assets )

assets created anew should be done taking into account the new gamma workflow and so they shouldn't display such errors.

In Nvidia cards and on windows that error also occurs.
If you deactivate HDR and shadowmaps (then you should have the old renderer and it should display fine, because the old renderer uses incorrect gamma math )

**Comment by shmerl** [comment]
> If you deactivate HDR and shadowmaps (then you should have the old renderer and it should display fine, because the old renderer uses incorrect gamma math )

How exactly it can be done? I suppose some entries in config file?

**Comment by BielBdeLuna** [comment]
```
r_useHDR 0
r_useShadowMapping 0
```

**Comment by shmerl** [comment]
Thanks, trying that now.

**Comment by shmerl** [comment]
Yep, that did it and it looks much better with game assets. Thanks for you help!

**Comment by shmerl** [comment]
Hm, I'm a bit confused now. I see that

set r_useHDR "0"
set r_useShadowMapping "0"

Are already in my  D3BFGConfig.cfg

And even if I remove them from there, they reappear. Are they default?

**Comment by BielBdeLuna** [comment]
both might be 1 by default

**Comment by shmerl** [comment]
Strange, somehow even when I remove those values from everywhere, they reappear in  D3BFGConfig.cfg on next restart with values 0.

**Comment by BielBdeLuna** [comment]
maybe you have to use the "+seta" as an argument after the game executable?

**Comment by shmerl** [comment]
No, no parameters passed. Also, I noticed that that gray haze isn't really gone when they are set to 0.

**Comment by BielBdeLuna** [comment]
you see that the whole screen turns darker when you go from HDR 1 to 0 isn't it?

**Comment by shmerl** [comment]
Yes, it does. I just tested in console setting it back to 1, and it turns brighter indeed.

**Comment by shmerl** [comment]
I suspect commands entered in console are somehow saved, and forced on next restart.

**Comment by BielBdeLuna** [comment]
then if you try it with the flashlight on you should see the difference in the light texture as well, do you see it?

**Comment by shmerl** [comment]
Without HDR the gray haze is less pronounced, but it's still there. With HDR it's very visible.

**Comment by shmerl** [comment]
And I can confirm, that console entered commands are preserved somehow, overriding what's set in config files even on next restart. That explains the confusion.

**Comment by BielBdeLuna** [comment]
the gray haze is there because the game developers wanted it there! it is there in vanilla doom3 flashlight also.

**Comment by shmerl** [comment]
Really? I didn't realize that. It's supposed to be some dust or something?

**Comment by BielBdeLuna** [comment]
yes, the effects of lights over dust particles (expressed the doom3 way)
so it's not an error! you could say that RBdoom3BFG displays the texture wrong but, the game developers wanted that texture there.

**Comment by shmerl** [comment]
Here is another example: https://youtu.be/nPBgQCrHJ-Y?t=508

No such pixelated haze there. I suppose something is wrong after all.

**Comment by shmerl** [comment]
Yep, I see what you mean. I suppose the haze looks bad because of antialisaing not working properly.

**Comment by shmerl** [comment]
I dug up a computer with Nvidia card, and run the game with the blob. Visuals are clearly better (settings are the same).

**Comment by shmerl** [comment]
Does this mod work on Linux by the way? http://www.moddb.com/mods/doom-3-bfg-hi-def

**Comment by dukenukemx** [comment]
Just tried dsalt's changes and the game works but I see what people are saying about that flash light, cause wow that's ugly.  set r_useHDR "0" does fix it, but it also makes the game dark as hell, so no point really using it.  

Been a number of years since I played Doom 3 but the texture look a big ugly for some reason.

**Comment by BielBdeLuna** [comment]
if it annoys you, and you can live without it you can skin it out with a custom skin and apply to it a nodraw texture

**Comment by shmerl** [comment]
Setting `set r_useHDR "0"` doesn't make it to dark, actually without it it looks too bright for me. May be play with gamma correction somehow?

**Comment by dukenukemx** [comment]
I left HDR on and just cranked up the brightness so I don't even need to use the flash light.  I just want 2004 quality textures, cause the textures in this game are just so ugly.  There is no way that Doom 3 looked like that from my memory, though in 2004 I was tweaking the shit out of that game.  I really want Doom 3 BFG Hi def 3.0 for Linux.

**Comment by shmerl** [comment]
I agree, something is seriously off with textures here. I've red somewhere that in general, Doom 3 BFG degraded textures in comparison with original Doom 3, so it's probably not specific to open engines, but problem with assets themselves.

**Comment by dukenukemx** [comment]
How did you install the mod?

**Comment by shmerl** [comment]
I just copied it into the game directory like they recommend. But there is probably something else I'm missing.

**Comment by dukenukemx** [comment]
Tried just copying the base directory and the game will load a save but then go back to the menu.  Something isn't working right.

**Comment by RobertBeckebans** [comment]
Doom 3 BFG did not ship with the uncompressed TGA textures but only with compressed DXT textures which are in the newly introduced id Tech 5 .bimage files. You can't have the same texture quality with them.
Even if you drop the original TGA files into the base folder then the engine will compress those textures into the DXT ones and use the compressed ones. Changing this system would break compatibility with the released assets.

For the flashlight: Well it is an issue with HDR and the flashlight. RBDOOM-3-BFG has an additional brightness pass with r_forceAmbient 0.2 by default so you don't need the flashlight that often.
I prefer the HDR mode because the colors are way more correct and natural than with the standard sRGB rendering.  See here: https://imgur.com/a/dKfLH

**Comment by shmerl** [comment]
> Changing this system would break compatibility with the released assets.

So how is that mod above supposed to work according to its authors?

**Comment by BielBdeLuna** [comment]
well, he is the author...

**Comment by shmerl** [comment]
@BielBdeLuna: who is?

**Comment by BielBdeLuna** [comment]
@RobertBeckebans ^^^ the one you asked it too

**Comment by shmerl** [comment]
Ah, great. @RobertBeckebans: So how are those high res textures supposed to work in that mod?

**Comment by dukenukemx** [comment]
I'm curious too cause when you download those textures it comes with RBDoom3BFG.

**Comment by BielBdeLuna** [comment]
you can drop them into the base folder, or as a mod in it's own folder, the textures should come with a text file explaining how to open them.

**Comment by shmerl** [comment]
Looks like the mod author's repo is [here](https://github.com/y2keeth/RBDOOM3BFG-HiDef-Mod).

I wonder if code is the same, or it's modified to support the mod.

**Comment by shmerl** [comment]
@dukenukemx: OK, I was able to activate better texures. First I made a directory in game installation path (let's say `highresmod`). Then I placed all files from `base` in the archive (and patches for it) into `highresmod`.

Then run the game like this:

    ./RBDoom3BFG +set fs_resourceLoadPriority 0 +set fs_game highresmod

I used original compiled engine, but I suppose the forked one has some tweaks to better support those textures. Original one worked OK, but I couldn't use old save - you need to start a new game apparently. And on start it takes some time to generate converted textures as well.

**Comment by y2keeth** [comment]
wow lol

**Comment by shmerl** [comment]
@y2keeth: Hi! :) Do your textures supposed to work with stock RBDOOM-3-BFG, or your fork is needed to use them properly?

**Comment by y2keeth** [comment]
suppose to work with any rbdoom3bfg,  my fork just has a few things like sodes4funs stencil shadows fix and self shadowing added

**Comment by y2keeth** [comment]
i dont know why you are having issues, sorry

**Comment by dukenukemx** [comment]
@shmerl I tried what you said.  In the .rbdoom3bfg folder I made a highresmod folder and copied the base from Doom3BFG and then copied the files from the base folder from the texture mod.  Ran your command and when the game loads it just crashes back to menu.  Not sure what I'm doing wrong.

**Comment by shmerl** [comment]
@dukenukemx: No, you don't need to copy files from base of Doom3BFG. Just copy files from the base of the mod.

**Comment by dukenukemx** [comment]
@shmerl Thanks I got it working.  I can tell that not all the textures were upgraded but it still looks a lot better than it did before.

**Comment by shmerl** [comment]
Yep, new game is required it seems.

**Comment by y2keeth** [comment]
this is why I miss the installer

**Comment by y2keeth** [comment]
still no idea what you are doing wrong.  should be just extract the 7z into the doom 3 bfg folder (where doom 3 bfg .exe is), thats it.

**Comment by Hi-Angel** [comment]
Hi, I occasionally got here from a forum. I took a cursory look through the thread, and I want to note for everyone that you can't blindly install latest development drivers, and go complain to peoples unrelated to drivers. Well, not unless you bugreport contains sentence `I also tried with Mesa stable, and I see the same issue`. My post is the first on this page containing the word `stable`, but I see words `Mesa 17.3.0-devel` in someone's version string.

Damn, you should be thankful your system is working at all! The feature freeze before releasing to stable doesn't happens for lols! Your algorithm to bugreporting should be:

	if (Mesa development is fine)
		return;
	else if (I know where to report because I have carefully investigated the issue)
		go report; // you're acknowledgeable of the problem background
	else {
		if (Mesa stable is fine)
			report to Mesa;
		else
			report to application developer;
	}

**Comment by shmerl** [comment]
Testing Mesa devel is perfectly fine.

**Comment by Hi-Angel** [comment]
I am not saying it's not, I am saying that if you see problems with devel, you should either provide technical details about why the problem is in the application, or to mention that you've tested stable too. OP did neither.

**Comment by terzag** [comment]
Hello,
I have the same "trashed graphics" issue with a RadeonHD 7770 and Mesa 17.3.1 (which is now the current stable version). Installed rbdoom3bfg from the Ubuntu 17.10 repos, which seems to be version 1.1.0~preview3 / GIT 19/10/2016.

As there been some progress on the matter?

**Comment by thedanish** [comment]
terzag, just compiled from git and it works fine. On Ubuntu 16.04 with Mesa 17.

This issue should be closed.

**Comment by andrewp707** [comment]
Hello, i am running Ubuntu 18.04 with nvidia drivers installed and working.
I have just installed RBDOOM-3-BFG from the Ubuntu repo in . deb format.
I copied over the data from my Windows install of GOG Doom 3 BFG.
The game can read my install and works but when the screen loads its all distorted.
It looks like this might be a mesa issue like what this thread is talking about.

Here is the output from my console:

While compiling vertex program renderprogs/AmbientOcclusion_minify.vertex
-----------------
  1: // filename renderprogs/AmbientOcclusion_minify.vertex
  2: #version 150
  3: #define PC
  4: 
  5: float saturate( float v ) { return clamp( v, 0.0, 1.0 ); }
  6: vec2 saturate( vec2 v ) { return clamp( v, 0.0, 1.0 ); }
  7: vec3 saturate( vec3 v ) { return clamp( v, 0.0, 1.0 ); }
  8: vec4 saturate( vec4 v ) { return clamp( v, 0.0, 1.0 ); }
  9: vec4 tex2Dlod( sampler2D sampler, vec4 texcoord ) { return textureLod( sampler, texcoord.xy, texcoord.w ); }
 10: 
 11: 
 12: 
 13: in vec4 in_Position;
 14: in vec2 in_TexCoord;
 15: 
 16: out vec2 vofi_TexCoord0;
 17: 
 18: void main() {
 19: 	gl_Position = in_Position ;
 20: 	vofi_TexCoord0 = in_TexCoord ;
 21: }
-----------------
0:2(10): error: GLSL 1.50 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.00 ES, 3.00 ES, 3.10 ES, and 3.20 ES


It keeps giving me similar errors over and over in the console.

Is this a problem with the game or my GLSL?
